dask/layers.py

```diff
@@ -100,6 +102,235 @@ def __init__(
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def is_materialized(self):
+        return hasattr(self, "_cached_dict")
+
+    def _dask_keys(self):
+        if self._cached_keys is not None:
+            return self._cached_keys
+
+        name, chunks, numblocks = self.name, self.chunks, self.numblocks
+
+        def keys(*args):
+            if not chunks:
+                return [(name,)]
+            ind = len(args)
+            if ind + 1 == len(numblocks):
+                result = [(name,) + args + (i,) for i in range(numblocks[ind])]
+            else:
+                result = [keys(*(args + (i,))) for i in range(numblocks[ind])]
+            return result
+
+        self._cached_keys = result = keys()
+        return result
+
+    def _construct_graph(self, deserializing=False):
+        """Construct graph for a simple overlap operation."""
+        axes = self.axes
+        chunks = self.chunks
+        name = self.name
+        dask_keys = self._dask_keys()
+
+        getitem_name = "getitem-" + self.token
+        overlap_name = "overlap-" + self.token
+
+        if deserializing:
+            # Use CallableLazyImport objects to avoid importing dataframe
+            # module on the scheduler
+            concatenate3 = CallableLazyImport("dask.array.core.concatenate3")
+        else:
+            # Not running on distributed scheduler - Use explicit functions
+            from dask.array.core import concatenate3
+
+        dims = list(map(len, chunks))
+        expand_key2 = toolz.partial(expand_key, dims=dims, axes=axes)
+
+        # Make keys for each of the surrounding sub-arrays
+        interior_keys = toolz.pipe(
+            dask_keys, flatten, map(expand_key2), map(flatten), toolz.concat, list
+        )
+        interior_slices = {}
+        overlap_blocks = {}
+        for k in interior_keys:
+            frac_slice = fractional_slice((name,) + k, axes)
+            if (name,) + k != frac_slice:
+                interior_slices[(getitem_name,) + k] = frac_slice
+            else:
+                interior_slices[(getitem_name,) + k] = (name,) + k
+                overlap_blocks[(overlap_name,) + k] = (
+                    concatenate3,
+                    (concrete, expand_key2((None,) + k, name=getitem_name)),
+                )
+
+        dsk = toolz.merge(interior_slices, overlap_blocks)
+        return dsk
+
+    @classmethod
+    def __dask_distributed_unpack__(cls, state):
+        return cls(**state)._construct_graph(deserializing=True)
+
+
+def expand_key(k, dims, name=None, axes=None):
```

Comment:
As a relative newcomer to dask, I often find the number of somewhat generically-named utility functions to be overwhelming. Since this function and the ones below it are being moved to a more central location (i.e., not namespaced by the `overlap` module), I think it would be nice to give them somewhat more specific names.

Perhaps we should also think about moving them into `dask.utils`, or a new `dask.layer_utils`.

Reply:
To be more concrete, maybe something like `_expand_keys_around_center`. Not sure about `fractional_slice` right now, but it seems very specific and could also likely be made semi-private.

Reply:
I've changed `expand_keys` to `_expand_keys_around_center`. Since we're not super sure about `fractional_slices`, let's have more of a think about that when we do the reorganization https://github.com/dask/dask/issues/7722 (I'm hoping that will help us clarify things)